File: 221116_스택/6198_옥상 정원 꾸미기/6198.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        long ans = 0;
        Stack<Integer> stack = new Stack<>();
        for (int i = 0; i < n; i++) {
            int input = Integer.parseInt(br.readLine());
            while (!stack.isEmpty() && stack.peek() <= input) {
                stack.pop();
            }
            stack.push(input);
            ans += stack.size() - 1;
        }
        System.out.print(ans);
    }
}

File: 221116_스택/6198_옥상 정원 꾸미기/6198.py
n = int(input())
arr = [int(input()) for _ in range(n)]

stack = []
ans = 0

for i in range(n - 1, -1, -1):
    cnt = 0
    while stack:
        if stack[-1][0] < arr[i]:
            cnt += stack[-1][1] + 1
            del stack[-1]
        else:
            stack.append([arr[i], cnt])
            break

    if not stack: stack.append([arr[i], cnt])
    ans += cnt
print(ans)
